lightning/src/blinded_path/utils.rs

```diff
@@ -177,3 +177,26 @@ impl Writeable for Padding {
 		Ok(())
 	}
 }
+
+
+/// A wrapper struct that stores the largest packet size for the given [`BlindedPath`].
+/// This helps us calculate the appropriate padding size for the tlvs when writing them.
+pub struct WithPadding<T: Writeable> {
+	/// Length of the packet with the largest size in the [`BlindedPath`].
+	pub(crate) max_length: usize,
+	/// The current packet's TLVs.
+	pub(crate) tlvs: T,
+}
+
+impl<T: Writeable> Writeable for WithPadding<T> {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		let length = self.max_length.saturating_sub(self.tlvs.serialized_length());
+		let padding = Some(Padding::new(length));
+
+		encode_tlv_stream!(writer, {
+			(1, padding, option)
+		});
+
+		self.tlvs.write(writer)
```

Comment:
While this works, if used incorrectly we could end up writing two TLVs with type 1. Let's implement a `padding` method to return `Padding` as calculated above. Then instead of implementing `Writeable` on `WithPadding<T>` here, in the next commit we can change the implementation to be on both `WithPadding<ForwardTlvs>` and `WithPadding<ReceiveTlvs>`, calling the new `padding` as needed. Then the use of `encode_tlv_stream` there could include the padding TLV along with the other ones when calling `encode_tlv_stream`. We could also mark the `padding` method `#[inline(always)]` to help the compiler remove any redundancy.

Reply:
Thanks a lot for the suggestion, Jeff!

I understand the concern about potentially ending up with two TLVs of type 1, which we want to avoid. However, I have a few doubts I’d love to clarify regarding the solution.

Since `WithPadding` needs the `T: Writeable` condition for the TLVs (because we’re using `serialise_length()`), we'd have to maintain `Writeables` for `ForwardTlvs` and `ReceiveTlvs`, while also introducing almost identical new `Writeables` with padding.

Additionally, we’d need to define `Writeable` for `WithPadding<ControlTlvs>`, as we use `ControlTlvs` within `blinded_hops`.

So, if I understand your suggestion correctly, we would need to introduce three new `Writeables` for the message itself.

Let me know what you think about it! Also, in case I misunderstood your suggestion, please feel free to correct me!

Thanks a lot, again!

Reply:
Hmm... right, good point. That's probably worse than what you already have. I think we'd have to forget about `WithPadding` and instead add an `Option<Padding>` field to `ForwardTlvs` and `ReceiveTlvs ` such that `None` is used when computing serialized length. Then it could be changed to `Some(Padding)` after computing the max length, for instance by having a `pad_to_length` method on `ControlTlvs`.

Here's the gist of it:

```diff
diff --git a/lightning/src/blinded_path/message.rs b/lightning/src/blinded_path/message.rs
index 36038a5c3..6de79f24a 100644
--- a/lightning/src/blinded_path/message.rs
+++ b/lightning/src/blinded_path/message.rs
@@ -17,7 +17,7 @@ use bitcoin::secp256k1::{self, PublicKey, Secp256k1, SecretKey};
 use crate::prelude::*;
 
 use crate::blinded_path::{BlindedHop, BlindedPath, IntroductionNode, NextMessageHop, NodeIdLookUp};
-use crate::blinded_path::utils::{self, WithPadding};
+use crate::blinded_path::utils::{self, Padding};
 use crate::io;
 use crate::io::Cursor;
 use crate::ln::channelmanager::PaymentId;
@@ -46,6 +46,7 @@ pub struct ForwardNode {
 /// route, they are encoded into [`BlindedHop::encrypted_payload`].
 #[derive(Clone)]
 pub(crate) struct ForwardTlvs {
+       pub(crate) padding: Option<Padding>,
        /// The next hop in the onion message's path.
        pub(crate) next_hop: NextMessageHop,
        /// Senders to a blinded path use this value to concatenate the route they find to the
@@ -56,6 +57,7 @@ pub(crate) struct ForwardTlvs {
 /// Similar to [`ForwardTlvs`], but these TLVs are for the final node.
 #[derive(Clone)]
 pub(crate) struct ReceiveTlvs {
+       pub(crate) padding: Option<Padding>,
        /// If `context` is `Some`, it is used to identify the blinded path that this onion message is
        /// sending to. This is useful for receivers to check that said blinded path is being used in
        /// the right context.
@@ -69,6 +71,7 @@ impl Writeable for ForwardTlvs {
                        NextMessageHop::ShortChannelId(scid) => (None, Some(scid)),
                };
                encode_tlv_stream!(writer, {
+                       (1, self.padding, option),
                        (2, short_channel_id, option),
                        (4, next_node_id, option),
                        (8, self.next_blinding_override, option)
@@ -80,6 +83,7 @@ impl Writeable for ForwardTlvs {
 impl Writeable for ReceiveTlvs {
        fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
                encode_tlv_stream!(writer, {
+                       (1, self.padding, option),
                        (65537, self.context, option),
                });
                Ok(())
@@ -184,17 +188,17 @@ pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
                        Some(scid) => NextMessageHop::ShortChannelId(scid),
                        None => NextMessageHop::NodeId(*pubkey),
                })
-               .map(|next_hop| ControlTlvs::Forward(ForwardTlvs { next_hop, next_blinding_override: None }))
-               .chain(core::iter::once(ControlTlvs::Receive(ReceiveTlvs{ context: Some(context) })));
+               .map(|next_hop| ControlTlvs::Forward(ForwardTlvs { padding: None, next_hop, next_blinding_override: None }))
+               .chain(core::iter::once(ControlTlvs::Receive(ReceiveTlvs{ padding: None, context: Some(context) })));
 
        let max_length = tlvs.clone()
                .map(|tlv| tlv.serialized_length())
                .max()
                .unwrap_or(0);
 
-       let length_tlvs = tlvs.map(|tlvs| WithPadding { max_length, tlvs });
+       let padded_tlvs = tlvs.map(|tlvs| tlvs.pad_to_length(max_length));
 
-       utils::construct_blinded_hops(secp_ctx, pks, length_tlvs, session_priv)
+       utils::construct_blinded_hops(secp_ctx, pks, padded_tlvs, session_priv)
 }
 
 // Advance the blinded onion message path by one hop, so make the second hop into the new
@@ -216,7 +220,7 @@ where
        let mut reader = FixedLengthReader::new(&mut s, encrypted_control_tlvs.len() as u64);
        match ChaChaPolyReadAdapter::read(&mut reader, rho) {
                Ok(ChaChaPolyReadAdapter {
-                       readable: ControlTlvs::Forward(ForwardTlvs { next_hop, next_blinding_override })
+                       readable: ControlTlvs::Forward(ForwardTlvs { padding: _, next_hop, next_blinding_override })
                }) => {
                        let next_node_id = match next_hop {
                                NextMessageHop::NodeId(pubkey) => pubkey,
```

With the new function taken from `WithPadding`.

```diff
diff --git a/lightning/src/onion_message/packet.rs b/lightning/src/onion_message/packet.rs
index 4f74c11fd..6c66706a0 100644
--- a/lightning/src/onion_message/packet.rs
+++ b/lightning/src/onion_message/packet.rs
@@ -314,16 +314,29 @@ pub(crate) enum ControlTlvs {
        Receive(ReceiveTlvs),
 }
 
+impl ControlTlvs {
+       pub fn pad_to_length(mut self, length: usize) -> Self {
+               let length = length.checked_sub(self.serialized_length());
+               debug_assert!(length.is_some(), "Size of this packet should not be larger than the size of largest packet.");
+               let padding = Some(Padding::new(length.unwrap()));
+               match &mut self {
+                       Self::Forward(tlvs) => tlvs.padding = padding,
+                       Self::Receive(tlvs) => tlvs.padding = padding,
+               }
+
+               self
+       }
+}
+
```

The change touches a few other places, but they are mostly minor. You can try this with a fixup commit and we can decide which approach is preferable.

Reply:
Wow! Thank you so much for the detailed description and the gist.
They were super helpful in updating the code with the new approach!

I've introduced the `Option<Padding>` approach as an `f:` commit in [pr3177.08](https://github.com/shaavan/rust-lightning/commits/pr3177.08). Thanks a lot!